pkg/collector/remoteconfig/remote_config.go

```diff
@@ -0,0 +1,114 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+// Package remoteconfig schedule and unschedule integrations through remote-config
+// using the AGENT_INTEGRATIONS product
+package remoteconfig
+
+import (
+	"encoding/json"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/secrets"
+	pkglog "github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// Scheduler is the structure used to run checks with RC
+type Scheduler struct {
+	scheduler     *collector.CheckScheduler
+	runningChecks []integration.Config
+}
+
+type agentIntegration struct {
+	Name       string            `json:"name"`
+	Instances  []json.RawMessage `json:"instances"`
+	InitConfig json.RawMessage   `json:"init_config"`
+}
+
+// secretsDecrypt allows tests to intercept calls to secrets.Decrypt.
+var secretsDecrypt = secrets.Decrypt
+
+// NewScheduler creates an instance of a remote config integration scheduler
+func NewScheduler() *Scheduler {
+	return &Scheduler{
+		runningChecks: make([]integration.Config, 0),
+	}
+}
+
+// AddScheduler initiate the remote-config scheduler
+func (sc *Scheduler) AddScheduler(scheduler *collector.CheckScheduler) {
+	if sc.scheduler != nil {
+		pkglog.Errorf("Remote-config scheduler is already initiated")
+	} else {
+		sc.scheduler = scheduler
+	}
+}
+
+// IntegrationScheduleCallback is called at every AGENT_INTEGRATIONS to schedule/unschedule integrations
+func (sc *Scheduler) IntegrationScheduleCallback(updates map[string]state.RawConfig, applyStateCallback func(string, state.ApplyStatus)) {
+	if sc.scheduler == nil {
+		pkglog.Debugf("Scheduler is not ready to start remote-config integrations")
+		return
+	}
+
+	// Unschedule every integrations, even if they haven't changed
+	sc.scheduler.Unschedule(sc.runningChecks)
```

Comment:
Is this safe to do in a separate go-routine? This will be running in the goroutine executing the remote config update processing.

Reply:
I did fully check, and the scheduling part is using locks under the hood https://github.com/DataDog/datadog-agent/blob/3ff5c025fb32e0d8c40be0f3a14e5706dfb6c676/pkg/collector/scheduler.go#L236-L259
and the collector part as well https://github.com/DataDog/datadog-agent/blob/3ff5c025fb32e0d8c40be0f3a14e5706dfb6c676/pkg/collector/collector.go#L117-L119

However, there is indeed one issue with the `configToChecks map[string][]checkid.ID` when unscheduling, so I added the use of the structure mutex, nice catch 👍